### PR TITLE
Add Performant WP Query Check

### LINKED ISSUE #27: Create Performant_WP_Query_Params_Check
## Description

Checks for the usage of non-performant `WP_Query` arguments. These will include sniffs that check and warn for use of arguments such as:

- `post__not_in`
- passing `posts_per_page => -1` or a high limit above a threshold
- passing `cache_results => false`.
- usage of `meta_query` and `term_query` keys.

This will use the existing `WordPressVIPMinimum.Performance.WPQueryParams` sniff in the [VIP Coding Standards](https://github.com/Automattic/VIP-Coding-Standards) and the `WordPress.DB.SlowDBQuery` sniff in the [WordPress Coding Standards](https://github.com/WordPress/WordPress-Coding-Standards/).

## Acceptance Criteria

- Class `Performant_WP_Query_Params_Check` should be created and exists at `includes/Checker/Checks`
- Class `Performant_WP_Query_Params_Check` should extend the `Abstract_PHP_CodeSniffer_Check` abstract class
- Class should implement the public `get_args()` and return an associative array of the required PHPCS arguments
    - `'sniffs' => 'WordPress.DB.SlowDBQuery, WordPressVIPMinimum.Performance.WPQueryParams'`
    - `'standard' => 'WordPress,WordPressVIPMinimum'`
- A new test plugin should be created at `tests/phpunit/testdata/plugins/test-plugin-performant-wp-query-params/`
    - Test plugin should contain code to trigger a warning in the check using the following code.

```
    $query = new WP_Query(
         array(
		'post__not_in'   => array( 1, 2, 3 ),
		'posts_per_page' => -1,
		'cache_results'  => false,
		'meta_key'       => 'age',
		'meta_query'     => array(
			array(
				'key'     => 'age',
				'value'   => array( 3, 4 ),
				'compare' => 'IN',
			),
		),
		'tax_query'      => array(
			array(
				'taxonomy' => 'custom_taxonomy_slug',
				'operator' => 'EXISTS',
			),
		),
	  )
    );
```

## Tests Coverage
- `run()` method adds warnings to the `$check_result` when checking an invalid file - check reference [here](https://github.com/10up/plugin-check/blob/trunk/tests/Checker/Checks/I18n_Usage_Check_Tests.php#L20)
- `run()` method does not add any warnings to the `$check_result` when checking a valid file - check reference [here](https://github.com/10up/plugin-check/blob/trunk/tests/Checker/Checks/I18n_Usage_Check_Tests.php#L51)

### PR DESCRIPTION
**Pull request by jjgrainger**
Closes #27

### CODE REVIEW

File: tests/phpunit/Checker/Checks/Performant_WP_Query_Params_Check_Tests.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+use WordPress\Plugin_Check\Checker\Check_Context;
+use WordPress\Plugin_Check\Checker\Check_Result;
+use WordPress\Plugin_Check\Checker\Checks\Performant_WP_Query_Params_Check;
+use WordPress\Plugin_Check\Test_Utils\TestCase\Static_Check_UnitTestCase;
+
+class Performant_WP_Query_Params_Check_Tests extends Static_Check_UnitTestCase {
+
+	public function test_run_with_errors() {
+		$performant_query = new Performant_WP_Query_Params_Check();
+		$check_context    = new Check_Context( TESTS_PLUGIN_DIR . '/tests/phpunit/testdata/plugins/test-plugin-performant-wp-query-params-errors/load.php' );
```
Comment (felixarntz) [inline tests/phpunit/Checker/Checks/Performant_WP_Query_Params_Check_Tests.php]: After merging `trunk` back in, please use the new constant here that was introduced in #189.

File: tests/phpunit/Checker/Checks/Performant_WP_Query_Params_Check_Tests.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		// Check for WordPress.DB.SlowDBQuery.slow_db_query_tax_query warning.
+		$this->assertArrayHasKey( 34, $warnings['load.php'] );
+		$this->assertArrayHasKey( 26, $warnings['load.php'][34] );
+		$this->assertArrayHasKey( 'code', $warnings['load.php'][34][26][0] );
+		$this->assertEquals( 'WordPress.DB.SlowDBQuery.slow_db_query_tax_query', $warnings['load.php'][34][26][0]['code'] );
+	}
+
+	public function test_run_without_errors() {
+		$performant_query = new Performant_WP_Query_Params_Check();
+		$check_context    = new Check_Context( TESTS_PLUGIN_DIR . '/tests/phpunit/testdata/plugins/test-plugin-performant-wp-query-params-without-errors/load.php' );
```
Comment (felixarntz) [inline tests/phpunit/Checker/Checks/Performant_WP_Query_Params_Check_Tests.php]: Same as above.

**Review by felixarntz — APPROVED** [review]
@jjgrainger LGTM! This just needs to have the merge conflict resolved, and related, the tiny comments below. Good to approve already though!

File: composer.json [hunk 3]
```diff
@@ -6,8 +6,9 @@
   "require": {
     "php": ">=5.6",
     "composer/installers": "^v1.12.0 || ^2.2",
-    "dealerdirect/phpcodesniffer-composer-installer": "^1.0.0",
-    "wp-coding-standards/wpcs": "dev-develop#fca9d9e"
+    "dealerdirect/phpcodesniffer-composer-installer": "^0.4.1 || ^0.5 || ^0.6.2 || ^0.7 || ^1.0",
+    "wp-coding-standards/wpcs": "^2.3 | dev-develop#fca9d9e",
```
Comment (mukeshpanchal27) [inline composer.json]: This changes introduce #136 issue again.

**Review by mukeshpanchal27 — CHANGES_REQUESTED** [review]
The changes made for Composer have reintroduced the PHP 8.1 issue

File: composer.json (hunk 3 above)
Comment (spacedmonkey) [reply composer.json]: @jjgrainger Can you explain why this is needed?

Comment (jjgrainger) [reply composer.json]: @spacedmonkey @mukeshpanchal27 this check uses a sniff from the VIP coding standards (`automattic/vipwpcs`).

Trying to add the package while using `"wp-coding-standards/wpcs": "dev-develop#fca9d9e"` caused issues with a number of dependencies. This resolved those dependency issues but now there is a regression with #136 which I'm looking into.

File: composer.json [hunk 4]
```diff
@@ -6,8 +6,9 @@
...
+    "automattic/vipwpcs": "^2.0.0"
```
Comment (spacedmonkey) [inline composer.json]:
```suggestion
    "dealerdirect/phpcodesniffer-composer-installer": "^1.0.0",
    "wp-coding-standards/wpcs": "dev-develop#fca9d9e as 2.3.0",
    "automattic/vipwpcs": "^2.0.0"
```

Using some functionality called [aliases](https://getcomposer.org/doc/articles/aliases.md) is a better way around this.

**Comment by spacedmonkey** [comment]
@jjgrainger @mukeshpanchal27 There seems to be a valid until test failure here. 

Also, this should fix the issues with composer. https://github.com/10up/plugin-check/pull/193#discussion_r1235184078

File: composer.json (hunk 3 above)
Comment (spacedmonkey) [reply composer.json]: Here is the fix https://github.com/10up/plugin-check/pull/193/files#r1235184078

Comment (jjgrainger) [reply composer.json]: Thanks @spacedmonkey the alias fixes the #136 regression, however I'm seeing a new issue.

Seems there is still some issues with dependencies as now this performant query check no longer works. When testing this directly with PHPCS I see a new `ERROR: Referenced sniff "WordPress.PHP.StrictComparisons" does not exist`.

It looks some sniffs have been moved in WPCS but not referenced correctly by the VIP package. I'm still investigating...

cc @mukeshpanchal27

Comment (spacedmonkey) [reply composer.json]: I can't replicate. Can you push up your branch @jjgrainger I can look at what you are seeing.

Comment (spacedmonkey) [reply composer.json]: Not related, but found in my search https://github.com/WordPress/WordPress-Coding-Standards/blob/f2b63051f08cf6d51360896bb1647eabc6911bf2/CHANGELOG.md?plain=1#L143

Comment (spacedmonkey) [reply composer.json]: This is my branch. https://github.com/10up/plugin-check/pull/196

Comment (jjgrainger) [reply composer.json]: Thanks @spacedmonkey 

Your solution works in order to pull in the correct dependencies, however I'm facing a new issue after this.

When I run the following command on my branch to make sure that PHPCS sniffs are working with the test plugin I get the follow error.

```
./vendor/bin/phpcs --standard="WordPress,WordPressVIPMinimum" --sniffs="WordPress.DB.SlowDBQuery,WordPressVIPMinimum.Performance.WPQueryParams" ./tests/phpunit/testdata/plugins/test-plugin-performant-wp-query-params-errors/load.php
```

```
ERROR: Referenced sniff "WordPress.PHP.StrictComparisons" does not exist
```

I believe this is to do with wpcs moving certain sniffs to a different name but the vipcs references the old names. I've tried to see if there is a combination of versions that will 1) resolve this issue and 2) solve the https://github.com/10up/plugin-check/issues/136 regression. However, I've had no luck.

As it looks like we can't get to an ideal solution, I believe the best option to move forward quickly is to remove the vipcs package and have this check only run the `WordPress.DB.SlowDBQuery` sniff for now.

We can then create a new issue to add the `WordPressVIPMinimum.Performance.WPQueryParams` sniff back in once both vipcs and wpcs packages have been updated and we can use compatible stable versions that do no reintroduce the issue with #136 

I'd like to hear your thoughts on this @mukeshpanchal27 @felixarntz @spacedmonkey and if agreed we can make those updates and finalise this branch.

Thanks!

Comment (mukeshpanchal27) [reply composer.json]: @jjgrainger We discuss this here: https://github.com/10up/plugin-check/pull/196#discussion_r1236471117

**Review by spacedmonkey — APPROVED** [review]

File: includes/Checker/Abstract_Check_Runner.php [hunk 5]
```diff
@@ -376,6 +376,7 @@ private function register_checks() {
 				'late_escaping'              => new Checks\Late_Escaping_Check(),
 				'plugin_updater'             => new Checks\Plugin_Updater_Check(),
 				'plugin_review_phpcs'        => new Checks\Plugin_Review_PHPCS_Check(),
+				'performant_wp_query_params' => new Checks\Late_Escaping_Check(),
```
Comment (mukeshpanchal27) [inline includes/Checker/Abstract_Check_Runner.php]:
```suggestion
				'performant_wp_query_params' => new Checks\Performant_WP_Query_Params_Check(),
```
Wrong class assign in https://github.com/10up/plugin-check/pull/193/commits/eeb27193ab1b3d130bfb659e9ffc78c5ed0afc9e

**Review by mukeshpanchal27 — APPROVED** [review]
Thanks @jjgrainger and @spacedmonkey great work.

**Pull request merged**